DigitalLearningSolutions.Web/Controllers/TrackingSystem/CentreConfiguration/RegistrationPromptsController.cs

```diff
@@ -71,32 +90,83 @@ public IActionResult EditRegistrationPrompt(EditRegistrationPromptViewModel mode
             {
                 SaveAction => EditRegistrationPromptPostSave(model),
                 AddPromptAction => RegistrationPromptAnswersPostAddPrompt(model),
+                BulkAction => EditRegistrationPromptBulk(model),
                 _ => RedirectToAction("Error", "LearningSolutions")
             };
         }
 
         [HttpGet]
-        [Route("Add/SelectPrompt")]
-        public IActionResult AddRegistrationPromptSelectPrompt()
+        [Route("{promptNumber:int}/Edit/Bulk")]
+        [ServiceFilter(typeof(RedirectEmptySessionData<EditRegistrationPromptData>))]
+        public IActionResult EditRegistrationPromptBulk(int promptNumber)
         {
-            var addRegistrationPromptData = TempData.Peek<AddRegistrationPromptData>();
+            var data = TempData.Peek<EditRegistrationPromptData>()!;
 
-            if (addRegistrationPromptData == null || !Request.Cookies.ContainsKey(CookieName))
+            var model = new BulkRegistrationPromptAnswersViewModel(
+                data.EditModel.OptionsString,
+                false,
+                promptNumber
+            );
+
+            return View("BulkRegistrationPromptAnswers", model);
+        }
+
+        [HttpPost]
+        [Route("Bulk")]
+        [ServiceFilter(typeof(RedirectEmptySessionDataXor<AddRegistrationPromptData, EditRegistrationPromptData>))]
+        public IActionResult RegistrationPromptBulkPost(BulkRegistrationPromptAnswersViewModel model)
+        {
+            ValidateBulkOptionsString(model.OptionsString);
+            if (!ModelState.IsValid)
             {
-                addRegistrationPromptData = new AddRegistrationPromptData();
-                var id = addRegistrationPromptData.Id;
-
-                Response.Cookies.Append(
-                    CookieName,
-                    id.ToString(),
-                    new CookieOptions
-                    {
-                        Expires = DateTimeOffset.UtcNow.AddDays(1)
-                    }
-                );
-                TempData.Set(addRegistrationPromptData);
+                return View("BulkRegistrationPromptAnswers", model);
+            }
+
+            if (model.IsAddPromptJourney)
+            {
+                var addData = TempData.Peek<AddRegistrationPromptData>()!;
+                addData.ConfigureAnswersViewModel!.OptionsString = model.OptionsString;
+                TempData.Set(addData);
+
+                return RedirectToAction("AddRegistrationPromptConfigureAnswers");
             }
 
+            var editData = TempData.Peek<EditRegistrationPromptData>()!;
+            editData.EditModel!.OptionsString = model.OptionsString;
+            TempData.Set(editData);
+
+            return RedirectToAction("EditRegistrationPrompt", new { promptNumber = model.PromptNumber });

```

Comment:
Now looking at this I wonder whether it would be worth creating two endpoints: `/Add/Bulk` and `/Edit/Bulk` but still use the same BulkRegistrationPromptAnswersViewModel. Given that you've pulled the ValidateBulkOptionsString(model.OptionsString); into a private method, the duplication between the methods would be minimised. What do you think?

Reply:
Can we not add the conditional for the asp-action rather than than the whole form tag?

Reply:
My only thinking there was we lose the validation that the endpoint exists if we do that.